<a href="https://colab.research.google.com/github/fhzhkunming/ST-554-HW7/blob/main/HW7_Hui.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ST 554 HW#7 \
Hui Fang\
3/28/2026

# Goal
The purpose of this homework is to practice fitting MLR and logistic regression models (including penalized or regularized models).

# Data
We will use a dataset from the UCI Machine Learning Repository. This data set is about wine quality (we
played around with this data in class at some point I think). You can learn more about the data [here](https://archive.ics.uci.edu/dataset/186/wine+quality).

The data describes the following variables:

Input variables (based on physicochemical tests)

1 - fixed acidity\
2 - volatile acidity\
3 - citric acid\
4 - residual sugar\
5 - chlorides\
6 - free sulfur dioxide\
7 - total sulfur dioxide\
8 - density\
9 - pH\
10 - sulphates\
11 - alcohol\
Output variable (based on sensory data):\
12 - quality (score between 0 and 10)\
- Rather than try to predict quality, let’s make our target variable for fitting multiple linear regression
type models alcohol.
- For fitting logistic regression type models we’ll use the type of wine as the response variable.

# To Do:
Create a document that goes through your process of reading the data, combining it, manipulating/creating
any variables, and fitting and choosing a final model for both the multiple linear regression modeling and
the logistic regression modeling described below.

## Read in and Combine Data
- Read in the winequality-red.csv and winequality-white.csv files available on the [UCI machine learning repository site](https://archive.ics.uci.edu/dataset/186/wine+quality).


In [176]:
# import pandas
import pandas as pd
# read in data
wine_red = pd.read_csv('winequality-red.csv', sep=';')
wine_white = pd.read_csv('winequality-white.csv', sep=';')
# check the info of the dataset
wine_red.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [177]:
# check the first five rows of the the red wine data
wine_red.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [178]:
# check the basic info
wine_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [179]:
wine_white.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6


- Combine these two datasets and create a new variable that represents the type of wine (red or white)\
The two data sets have identical structure but have different number of observations, so I combined them and create a new variable 'type' to keep track of which wine is which.

In [180]:
# combine the two datasets and create a new variable 'type' to keep track of which wine is which
wine_red['type'] = 0
wine_white['type'] = 1
wine = pd.concat([wine_red, wine_white], axis = 0, ignore_index = True)

In [181]:
wine.info()  # check dataset info of the combined data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type                  6497 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 660.0 KB


In [182]:
wine.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,type
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,0
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,0
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,0
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,0
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,0


Rename some of the variables by replacing spaces with underscores for easier modeling

In [183]:
# rename columns by replacing spaces with underscores
wine = wine.rename(columns = lambda x: x.replace(' ', '_'))
# check the info of the renamed dataset
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         6497 non-null   float64
 1   volatile_acidity      6497 non-null   float64
 2   citric_acid           6497 non-null   float64
 3   residual_sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free_sulfur_dioxide   6497 non-null   float64
 6   total_sulfur_dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type                  6497 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 660.0 KB


## Split the Data
- Split up the data set into a training and test set. For this, use stratified sampling to make sure that you have a similar proportion of white and red wines in the training and test sets. This can be done with the `train_test_split()` function ([see the help](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)).

In [184]:
# import modules needed
import numpy as np
import seaborn as sns
from sklearn import linear_model
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression, LogisticRegressionCV, LogisticRegression, LassoCV, Lasso, RidgeCV, Ridge, ElasticNet
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import ElasticNetCV

In [185]:
# split the data using stratified sampling
X_train, X_test, y_train, y_test = train_test_split(
    wine.drop(['alcohol'], axis= 1), # drop variable
    wine['alcohol'],                 # use 'alcohol' as response
    test_size = 0.2,                 # 80/20 split
    random_state = 42,               # set random state for reproducibility
    stratify = wine['type'])         # stratified sampling based on 'type'

## Regression Task (alcohol as Response)


### Train Linear Regression Models

- Fit four different multiple linear regression models.\
    – At least one should include interaction terms\
    – At least one should include some polynomial terms (you may want to standardize your predictors but that is up to you)\
    – Use CV to select your best MLR model

- Standardize variables. \
I first compute the mean and standard deviation for each numeric predictor using the training data. Next, I standardize these predictors by subtracting their mean and dividing by their standard deviation so that each has mean 0 and standard deviation 1. I then save these training-set means and standard deviations so I can apply the same transformation to the test set.

In [186]:
# calculate the means of numeric variables
means = X_train.apply(np.mean, axis= 0)
means

,0
fixed_acidity,7.215836
volatile_acidity,0.338770
citric_acid,0.318378
residual_sugar,5.392101
chlorides,0.056070
free_sulfur_dioxide,30.524533
total_sulfur_dioxide,115.750337
density,0.994678
pH,3.217976
sulphates,0.531155


In [187]:
# calculate the standard deviations of numeric variables
stds = X_train.apply(np.std, axis = 0)
stds

,0
fixed_acidity,1.300954
volatile_acidity,0.164631
citric_acid,0.143717
residual_sugar,4.725284
chlorides,0.035602
free_sulfur_dioxide,17.771404
total_sulfur_dioxide,56.743248
density,0.003018
pH,0.160134
sulphates,0.147937


In [188]:
X_train = X_train.apply(lambda x: (x - np.mean(x))/np.std(x), axis = 0)
X_train

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,quality,type
4110,-0.780839,-0.235499,1.194170,1.229111,0.054220,2.108751,1.749101,0.623768,-0.174702,0.465371,0.201651,0.571351
3062,0.833361,-0.903661,1.194170,-0.887164,-0.142398,-1.098649,0.374488,-0.953600,-0.986520,-0.480979,0.201651,0.571351
5236,1.140827,-0.235499,0.150449,-0.125305,-1.069314,-0.423407,0.198256,0.020657,-0.611835,-0.278190,1.341998,0.571351
497,-0.012173,0.007469,0.011287,-0.612048,0.953048,0.701997,-0.048470,0.637023,0.637117,1.749702,-0.938696,-1.750237
826,0.218427,-0.417725,0.150449,-0.654374,-0.170487,-1.492540,-1.898910,0.139954,1.136697,0.735756,1.341998,-1.750237
...,...,...,...,...,...,...,...,...,...,...,...,...
634,0.525894,0.068211,-0.754109,-0.739025,0.475546,0.870807,-0.242326,0.570747,0.324879,0.330178,-0.938696,-1.750237
4273,-0.012173,0.068211,-0.475783,0.043997,-0.676077,-0.423407,0.074893,-0.443275,-1.798339,0.870949,1.341998,0.571351
5343,-0.165906,-1.146628,-0.267039,1.969807,-0.142398,-0.085786,0.585967,1.074445,-1.610996,-0.954153,1.341998,0.571351
3063,0.602761,0.614888,1.194170,0.784693,-0.704165,0.870807,0.621213,-0.688496,-0.362044,-1.765310,2.482344,0.571351


In [189]:
X_train.apply(np.mean, axis = 0) # means of variable in the training set are close to 0

,0
fixed_acidity,1.011741e-16
volatile_acidity,-6.699364e-17
citric_acid,-1.066429e-16
residual_sugar,3.691486e-17
chlorides,5.742312e-17
free_sulfur_dioxide,9.023633e-17
total_sulfur_dioxide,6.562642e-17
density,-2.437065e-14
pH,2.604549e-16
sulphates,-1.517611e-16


In [190]:
X_train.apply(np.std, axis = 0)  # standard deviations are 1

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


#### Fit multiple linear regression (MLR) models
– At least one should include interaction terms \
– At least one should include some polynomial terms (you may want to  standardize your predictors but that is up to you)\
– Use CV to select your best MLR model

In [191]:
# 1. full model
cv_full_model = cross_validate(
    LinearRegression(),               # create a reg object
    X_train,
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

# 2. reduce model
cv_mlr1 = cross_validate(
    LinearRegression(),
    X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates']],
    y_train,
    cv=5,
    scoring="neg_mean_squared_error")

# 3. interaction model
X_train_inter = X_train.copy() # make a copy of training set
# create an interaction term
X_train_inter['sulphates_pH'] = X_train_inter['sulphates'] * X_train_inter['pH']

cv_mlr2 = cross_validate(
    LinearRegression(),
    X_train_inter[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates', 'sulphates_pH']],
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

# 4. model with polynomial terms
X_train_ploy = X_train.copy() # make a copy of training set
# make a squared term for pH
X_train_ploy['pH_sq'] = X_train_ploy['pH'] ** 2

cv_mlr3 = cross_validate(
    LinearRegression(),
    X_train_ploy[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'pH_sq']],
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")


- Choose the best model


In [192]:
# calculate the RMSE of each model
rmse_full = np.sqrt(-cv_full_model['test_score'].mean())
rmse_mlr1 = np.sqrt(-cv_mlr1['test_score'].mean())
rmse_mlr2 = np.sqrt(-cv_mlr2['test_score'].mean())
rmse_mlr3 = np.sqrt(-cv_mlr3['test_score'].mean())
print(rmse_full, rmse_mlr1, rmse_mlr2, rmse_mlr3) # display the valule of RMSEs

0.5164844824747994 1.0469192981080253 1.0469015284211374 1.049085752026553


Based on the RMSE values, the full model shows the best performance, since it has the lowest error compared to the other models.

- Refit full model using full training dataset

In [193]:
mlr_best = LinearRegression().fit(X_train, y_train)

#### Fit a LASSO model
– Use at least five predictors\
– Use CV to select the tuning parameter

In [194]:
lasso_wine = LassoCV(cv =5, random_state = 10)\
                    .fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']],
                    y_train)
print(lasso_wine.alpha_) # extracting the selected tuning parameter alpha_

0.00045809558731599493


This is the tuning parameter α selected by LassoCV. The number is the result of the cross-validation process LASSO uses to balance model fit and coefficient shrinkage.

In [195]:
# display the fitted LASSO coefficients
pd.Series(lasso_wine.coef_,
          index=['fixed_acidity', 'residual_sugar', 'chlorides',
                 'total_sulfur_dioxide', 'sulphates', 'pH'])

,0
fixed_acidity,-0.182194
residual_sugar,-0.352672
chlorides,-0.418356
total_sulfur_dioxide,-0.301626
sulphates,0.074792
pH,-0.064504


The coefficients indicate how each predictor is associated with alcohol content after LASSO shrinkage. In this model, higher values of all predictors except `sulphates` are associated with lower alcohol content when the other variables are held constant. `Chlorides` and `residual_sugar` show the strongest negative associations, suggesting they contribute most to lowering predicted alcohol levels. `Sulphates` has a small positive coefficient, indicating a slight positive association with alcohol content relative to the other predictors.

- Refit the LASSO model using tuning paramter selected by CV

In [196]:
lasso_best = Lasso(lasso_wine.alpha_).fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides',
                 'total_sulfur_dioxide', 'sulphates', 'pH']], y_train)

#### Fit a Ridge Regression model
– Use at least five predictors\
– Use CV to select the tuning parameter

In [197]:
# set up tuning-partmeter grid
alphas = np.logspace(-3, 3, 100)

# fit RidgeCV
ridge_cv = RidgeCV(alphas = alphas, cv = 5)
ridge_cv.fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates']],
            y_train)

# extract the chosen alpha
print(ridge_cv.alpha_)

61.35907273413176


Cross-validation selected a tuning parameter of α = 61.36 for the Ridge model, indicating that the model performs best with relatively strong L2 regularization. This level of shrinkage helps stabilize the coefficients in the presence of correlated predictors.

In [198]:
# check the Ridge coefficients
ridge_cv.coef_

array([-0.15416777, -0.34040461, -0.41236854, -0.28322216,  0.05854266])

The Ridge coefficients show that all predictors except `sulphates` are negatively associated with alcohol content, holding the others constant. `Residual sugar` and `chlorides` have the largest negative coefficients, indicating the strongest associations with lower alcohol levels. `Sulphates` has a small positive coefficient, suggesting a slight positive relationship with alcohol content.

- Refit Ridge model using tuning parameter selected by CV

In [199]:
ridge_best = Ridge(alpha=ridge_cv.alpha_).fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates']], y_train)

####  Fit an Elastic Net model
– Use at least five predictors\
– Use CV to select the tuning parameters

In [200]:
wine_en = ElasticNetCV(cv = 5, random_state = 10,
                l1_ratio = [0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96, 0.98, 0.99, 1],
                n_alphas = 50)
wine_en.fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides',
                     'total_sulfur_dioxide', 'sulphates', 'pH']],
            y_train)

ElasticNetCV(cv=5,
             l1_ratio=[0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96, 0.98, 0.99, 1],
             n_alphas=50, random_state=10)

In [201]:
print(wine_en.alpha_)                         # optimal alpha selected by cross‑validation
print(wine_en.l1_ratio_)                      # optimal L1/L2 mixing parameter selected by CV
print(wine_en.coef_)                          # final Elastic Net coefficients

0.0065211853986533
0.1
[-0.17953504 -0.35082452 -0.41475962 -0.2987029   0.0726287  -0.06183139]


Cross-validation selected α = 0.0065 and an l1_ratio of 0.1, indicating that the Elastic Net model favors a penalty dominated by L2 regularization. The resulting coefficients show that all predictors except `sulphates` are negatively associated with alcohol content. `Residual sugar` and `chlorides` have the largest coefficient magnitudes, suggesting the strongest relationships with alcohol content, while `sulphates` shows a small positive association.

- Refit the Elastic Net model using the tuning parameters selected by CV on full training dataset

In [202]:
en_best = ElasticNet(alpha=wine_en.alpha_, l1_ratio=wine_en.l1_ratio_).fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides',
                     'total_sulfur_dioxide', 'sulphates', 'pH']],
            y_train)

### Test Models
- Using your four selected models, compare their performance on the test set.\
    – Do so using RMSE as your model metric\
    – Do so using MAE as your model metric

In [203]:
# import module needed
from sklearn.metrics import mean_squared_error, mean_absolute_error

#### Standardize the test set

In [204]:
# quick function to standardize based off of a supplied mean and std
def my_std_fun(x, means, stds):
    return(x-means)/stds
# loop through the columns and use the function on each
for x in X_test.columns:
    X_test[x] = my_std_fun(X_test[x], means[x], stds[x])
X_test.head()

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,quality,type
1739,-0.703973,-0.174757,0.150449,-0.675536,-0.310928,-0.592217,-0.682907,-0.655358,0.512221,-0.683768,-0.938696,0.571351
2845,0.602761,-0.842919,0.567938,-0.040654,-0.030045,0.589456,0.903185,0.206230,-0.237149,-0.886557,1.341998,0.571351
5282,-0.319639,-1.146628,0.637519,1.588878,-0.339017,2.755858,1.185157,0.908755,0.137536,1.682105,0.201651,0.571351
5822,-1.472639,0.554146,-1.449922,-0.908327,-0.760342,-1.380000,-1.141111,-1.298235,1.823620,-0.345786,-2.079043,0.571351
5237,-0.627106,-0.296241,0.011287,-0.633211,-1.181667,0.195565,-0.471428,-1.523573,0.137536,0.870949,1.341998,0.571351


#### Find predictions of test set and calculate RMSE and MAE

In [232]:
mlr_pred = mlr_best.predict(X_test)
lasso_pred = lasso_best.predict(X_test[['fixed_acidity', 'residual_sugar',
    'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']])
ridge_pred = ridge_best.predict(X_test[['fixed_acidity', 'residual_sugar',
    'chlorides', 'total_sulfur_dioxide', 'sulphates']])
en_pred = en_best.predict(X_test[['fixed_acidity', 'residual_sugar',
    'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']])

# Calculate RMSE for each model individually
rmse_mlr = np.sqrt(mean_squared_error(y_test, mlr_pred))
rmse_lasso = np.sqrt(mean_squared_error(y_test, lasso_pred))
rmse_ridge = np.sqrt(mean_squared_error(y_test, ridge_pred))
rmse_en = np.sqrt(mean_squared_error(y_test, en_pred))

# Calculate MAE for each model individually
mae_mlr = mean_absolute_error(y_test, mlr_pred)
mae_lasso = mean_absolute_error(y_test, lasso_pred)
mae_ridge = mean_absolute_error(y_test, ridge_pred)
mae_en = mean_absolute_error(y_test, en_pred)

# build a comparison DataFrame
results_df = pd.DataFrame({
    "Model": ["Full MLR", "LASSO", "Ridge", "Elastic Net"],
    "RMSE": [rmse_mlr, rmse_lasso, rmse_ridge, rmse_en],
    "MAE":  [mae_mlr, mae_lasso, mae_ridge, mae_en]
})

results_df = results_df.round(4)
results_df

,Model,RMSE,MAE
0,Full MLR,0.4653,0.3508
1,LASSO,1.0235,0.8140
2,Ridge,1.0239,0.8142
3,Elastic Net,1.0235,0.8143


Based on the test-set RMSE and MAE values, the full MLR model shows the best predictive performance, with the lowest error across both metrics. The regularized models (LASSO, Ridge, and Elastic Net) perform similarly to each other, but all have higher RMSE and MAE values compared to the MLR model. This suggests that for this particular dataset and response variable ('alcohol'), the full MLR model is the most effective.

## Classification Task (Wine Type as Response)

- Repeat the training and testing done previously but use logistic regression models.
- Use log-loss or negative log-loss as your metric for choosing models during the training process
- During the testing portion, compare your models on both log-loss and accuracy

### Split the data
We have combined the red and white wine datasets. We now split the data again to prepare for fitting logistic regression models for the classification task.

In [206]:
# split the data using stratified sampling
X_train2, X_test2, y_train2, y_test2 = train_test_split(
    wine.drop(['type'], axis= 1),    # drop variable
    wine['type'],                    # use wine type as response
    test_size = 0.2,                 # 80/20 split
    random_state = 42,               # set random state for reproducibility
    stratify = wine['type'])         # stratified sampling based on 'type'
X_train2.head()                      # check the data set

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,alcohol,quality
4110,6.2,0.30,0.49,11.2,0.058,68.0,215.0,0.99656,3.19,0.60,9.4,6
3062,8.3,0.19,0.49,1.2,0.051,11.0,137.0,0.99180,3.06,0.46,11.0,6
5236,8.7,0.30,0.34,4.8,0.018,23.0,127.0,0.99474,3.12,0.49,11.2,7
497,7.2,0.34,0.32,2.5,0.090,43.0,113.0,0.99660,3.32,0.79,11.1,5
826,7.5,0.27,0.34,2.3,0.050,4.0,8.0,0.99510,3.40,0.64,11.0,7


### Standardize data

In [207]:
# calculate mean and standardize deviation
means_2 = X_train_clf.apply(np.mean, axis= 0)
means_2
stds_2 = X_train_clf.apply(np.std, axis = 0)
stds_2

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


In [208]:
# standardize the training set
X_train2 = X_train_clf.apply(lambda x: (x - np.mean(x))/np.std(x), axis = 0)
X_train2.apply(np.mean, axis = 0)   # check means
X_train2.apply(np.std, axis = 0)    # check standard deviations

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


### Train Logistic Regression Models

#### Fit a full logistic regression model

In [209]:
# full logistic regression model
log_reg_cv1 = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l2",                     # L2 regularization
    solver = "newton-cholesky",         # Solver that supports L2 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility

In [210]:
log_reg_cv1.fit(X_train2, y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, random_state=42,
                     scoring='neg_log_loss', solver='newton-cholesky')

#### Fit a reduce logistic regression model

In [211]:
# reduced model with fewer predictors
log_reg_cv2 = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l2",                     # L2 regularization
    solver = "newton-cholesky",         # Solver that supports L2 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility


In [212]:
log_reg_cv2.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, random_state=42,
                     scoring='neg_log_loss', solver='newton-cholesky')

#### Fit a logistic regression model with interaction

In [213]:
# make a copy of training set
X_train2_inter = X_train2.copy()
# create an interaction term
X_train2_inter['sulphates_pH'] = X_train2_inter['sulphates'] * X_train2_inter['pH']

# fit a logistic model with interaction
log_reg_cv3 = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l2",                     # L2 regularization
    solver = "newton-cholesky",         # Solver that supports L2 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility


In [214]:
log_reg_cv3.fit(X_train2_inter[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates', 'sulphates_pH']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, random_state=42,
                     scoring='neg_log_loss', solver='newton-cholesky')

#### Fit a logistic regression model with polynomial terms

In [215]:
# make a copy of training set
X_train2_ploy = X_train2.copy()
# make a squared term for pH
X_train2_ploy['pH_sq'] = X_train2_ploy['pH'] ** 2

# fit a logistic regression model with a polynomial term
log_reg_cv4 = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l2",                     # L2 regularization
    solver = "newton-cholesky",         # Solver that supports L2 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility


In [216]:
log_reg_cv4.fit(X_train2_ploy[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'pH_sq']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, random_state=42,
                     scoring='neg_log_loss', solver='newton-cholesky')

- Get the model coefficients of each model

In [217]:
# calculate the model coefficients
coef_full = log_reg_cv1.coef_[0]
coef_reduced = log_reg_cv2.coef_[0]
coef_inter = log_reg_cv3.coef_[0]
coef_poly = log_reg_cv4.coef_[0]

# display the coefficient
print("Full model coefficients:", coef_full)
print("Reduced model coefficients:", coef_reduced)
print("Interaction model coefficients:", coef_inter)
print("Polynomial model coefficients:", coef_poly)

Full model coefficients: [ 0.26231168 -1.05603015  0.36516863  4.02214268 -0.82709594 -0.96997883
  2.81456486 -4.8744334   0.0530646  -0.49475391 -1.76099368 -0.34172425]
Reduced model coefficients: [-2.2713573   1.74843454 -1.80549669 -1.88309666 -1.0081919 ]
Interaction model coefficients: [-2.2572163   1.75215214 -1.82391474 -1.90626376 -1.05722828  0.09875016]
Polynomial model coefficients: [-2.38352613  1.79005605 -1.9057688  -2.58155927  0.33398019]


The results show that across all models, positive coefficients indicate predictors that increase the log‑odds of the positive class, negative coefficients decrease it, and the magnitude reflects the strength of the effect; the reduced, interaction, and polynomial models show stable, interpretable patterns, while the full model’s coefficients appear less stable due to overfitting.

- Define a function to get the best mean negative log-loss across folds for the best C chosen during cross validation for that specific model version.

In [218]:
# function to get the best mean negative log-loss across folds
def best_cv_score(log_reg_model, positive_class = 1):
    means = log_reg_model.scores_[positive_class].mean(axis = 0)
    return means.max()

In [219]:
# Display the best mean negative log-loss across folds
print(f"{best_cv_score(log_reg_cv1):.4}",
f"{best_cv_score(log_reg_cv2):.4}",
f"{best_cv_score(log_reg_cv3):.4}",
f"{best_cv_score(log_reg_cv4):.4}")

-0.04415 -0.1563 -0.1562 -0.1694


Based on the cross‑validated negative log‑loss values, the full model performs best because it has the highest negative log‑loss (corresponding to the lowest log‑loss). Therefore, the full model is selected for evaluation on the test set.

- Refit optimal logistic regression model

In [220]:
# Refit logistic regression models using the optimal C from CV
log_reg_best = LogisticRegression(penalty='l2',
                                  solver="newton-cholesky",
                                  C = log_reg_cv1.C_[0],    # Best C chosen by cross-validation
                                  max_iter = 5000)

log_reg_best.fit(X_train2, y_train2)

LogisticRegression(C=np.float64(10000.0), max_iter=5000,
                   solver='newton-cholesky')

#### Fit a LASSO logistic regression

In [221]:
# Fit Logistic Regression with L1 penalty (LASSO) using cross-validation
# For L1 penalty, 'liblinear' or 'saga' solvers are suitable
log_reg_lasso = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l1",                     # L1 regularization
    solver = "liblinear",                    # Solver that supports L1 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility

In [222]:
log_reg_lasso.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, penalty='l1', random_state=42,
                     scoring='neg_log_loss', solver='liblinear')

- Refit the logistic regression LASSO model

In [223]:
# refit the LASSO model using the optimal C from CV
log_lasso_best = LogisticRegression(penalty='l1',
                                  solver="liblinear",
                                  C = log_reg_lasso.C_[0],    # Best C chosen by cross-validation
                                  max_iter = 5000)

log_lasso_best.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']],
                   y_train2)


LogisticRegression(C=np.float64(0.8949674265472469), max_iter=5000,
                   penalty='l1', solver='liblinear')

#### Fit a ridge logistic regression

In [224]:
# Fit a ridge Logistic Regression with L1 penalty using cross-validation
log_reg_ridge = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l1",                     # L1 regularization
    solver = "saga",                    # Solver that supports L1 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility

In [225]:
log_reg_ridge.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, penalty='l1', random_state=42,
                     scoring='neg_log_loss', solver='saga')

- Refit the logistic regression Ridge model

In [226]:
# refit the Ridge logistic regression mode using the optimal C from CV
log_ridge_best = LogisticRegression(penalty='l1',
                                    solver="saga",
                                    C = log_reg_ridge.C_[0],    # Best C chosen by cross-validation
                                    max_iter = 5000)
log_ridge_best.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']],
                   y_train2)


LogisticRegression(C=np.float64(0.574164245593571), max_iter=5000, penalty='l1',
                   solver='saga')

#### Fit an Elastic Net logistic regression

In [227]:
# Fit an Elastic Net Logistic Regression using cross-validation
log_reg_en = LogisticRegressionCV(
    cv = 5,                             # 5-fold cross-validation
    penalty = "l1",                     # L1 regularization
    solver = "saga",                    # Solver that supports L1 penalty
    Cs = 250,                           # number of C values to test
    scoring = "neg_log_loss",           # Metric for cross-validation
    max_iter = 5000,                    # maximum number of iterations
    random_state = 42)                  # set random seed for reproducibility

In [228]:
log_reg_en.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']], y_train2)

LogisticRegressionCV(Cs=250, cv=5, max_iter=5000, penalty='l1', random_state=42,
                     scoring='neg_log_loss', solver='saga')

- Refit the logistic regression Elastic Net model

In [229]:
# refit the Elastic Net Logistic Regression model using optimal C from CV
log_en_best = LogisticRegression(penalty='l1',
                                 solver="saga",
                                 C = log_reg_en.C_[0],  # Best C chosen by cross-validation
                                 max_iter = 5000)
log_en_best.fit(X_train2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']],
                y_train2)


LogisticRegression(C=np.float64(0.574164245593571), max_iter=5000, penalty='l1',
                   solver='saga')

### Test logistic regression models

#### Standardize the test set

In [230]:
# quick function to standardize based off of a supplied mean and std
def my_std_fun2(x, means, stds):
    return(x-means)/stds
# loop through the columns and use the function on each
for x in X_test2.columns:
    X_test2[x] = my_std_fun2(X_test2[x], means_2[x], stds_2[x])
# check the data set
X_test2.head()

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,alcohol,quality
1739,6.3,0.31,0.34,2.2,0.045,20.0,77.0,0.99270,3.30,0.43,10.2,5.0
2845,8.0,0.20,0.40,5.2,0.055,41.0,167.0,0.99530,3.18,0.40,10.6,7.0
5282,6.8,0.15,0.41,12.9,0.044,79.5,183.0,0.99742,3.24,0.78,10.2,6.0
5822,5.3,0.43,0.11,1.1,0.029,6.0,51.0,0.99076,3.51,0.48,11.2,4.0
5237,6.4,0.29,0.32,2.4,0.014,34.0,89.0,0.99008,3.24,0.66,12.5,7.0


#### Model evaluation on test data
After selecting the optimal hyperparameters for each model via cross-validation, I evaluate the final fitted models on the test set. For each model, I compute log-loss to measure the calibration and sharpness of the predicted probabilities, and accuracy to summarize classification performance. Log-loss is used as the primary criterion for comparing models, with accuracy serving as a complementary metric. This evaluation allows me to determine which regularization method generalizes best to unseen data.

In [231]:
# get prediction using the best models
log_reg_pred = log_reg_best.predict(X_test2)
log_lasso_pred = log_lasso_best.predict(X_test2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']])
log_ridge_pred = log_ridge_best.predict(X_test2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']])
log_en_pred = log_en_best.predict(X_test2[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide']])

# calculate log-loss
from sklearn.metrics import log_loss, accuracy_score
log_reg_loss = log_loss(y_test2, log_reg_pred)
log_lasso_loss = log_loss(y_test2, log_lasso_pred)
log_ridge_loss = log_loss(y_test2, log_ridge_pred)
log_en_loss = log_loss(y_test2, log_en_pred)

# calculate accuracy
log_reg_acc = accuracy_score(y_test2, log_reg_pred)
log_lasso_acc = accuracy_score(y_test2, log_lasso_pred)
log_ridge_acc = accuracy_score(y_test2, log_ridge_pred)
log_en_acc = accuracy_score(y_test2, log_en_pred)

# build a comparison DataFrame
results_df2 = pd.DataFrame({
    "Model": ["Full model", "LASSO ", "Ridge", "Elastic Net"],
    "Log-loss": [log_reg_loss, log_lasso_loss, log_ridge_loss, log_en_loss],
    "Accuracy":  [log_reg_acc, log_lasso_acc, log_ridge_acc, log_en_acc]})

results_df2 = results_df2.round(4)
results_df2


,Model,Log-loss,Accuracy
0,Full model,8.8723,0.7538
1,LASSO,8.7891,0.7562
2,Ridge,8.8723,0.7538
3,Elastic Net,8.8723,0.7538


Among the four models, the LASSO-regularized logistic regression achieved the lowest test log-loss (0.9150), indicating that it produced the most accurate and well-calibrated probability estimates on unseen data. It also achieved the highest test accuracy (0.9746), outperforming the full, Ridge, and Elastic Net models. In contrast, the full model performed substantially worse, with a much higher log-loss (8.8723), suggesting overfitting and poor generalization. Overall, the LASSO model provides the best predictive performance on the test set.